pi-hole/pi-hole
# API and web integration

**By submitting this pull request, I confirm the following:**
- [x] I have read and understood the [contributors guide](https://github.com/pi-hole/pi-hole/blob/master/CONTRIBUTING.md), as well as this entire template.
- [ ] I have made only one major change in my proposed changes.
- [x] I have commented my proposed changes within the code.
- [x] I have tested my proposed changes, and have included unit tests where possible.
- [x] I am willing to help maintain this change if there are issues with it later.
- [x] I give this submission freely and claim no ownership.
- [x] It is compatible with the [EUPL 1.2 license](https://opensource.org/licenses/EUPL-1.1)
- [ ] I have squashed any insignificant commits. ([`git rebase`](http://gitready.com/advanced/2009/02/10/squashing-commits-with-rebase.html))

---

Note that this is a draft PR for now. It may stay this way until after the beta period is finished.

**How to test**
1. Start a fresh server
2. Clone this repository to `/etc/.pihole`
3. Checkout `feature/api`
4. Make a file at `/etc/pihole/ftlbranch` with the contents: `development`
5. Make a file at `/etc/pihole/apibranch` with the contents: `development`
6. Run `./automated\ install/basic-install.sh`

When the installer is finished, you will have the web interface running on port 80. The API will be able to connect to FTL and serve stats to the web interface. All web interface functions should work, including changing DNS options (it will reflect in `/etc/dnsmasq.d/pihole.conf`).

**What does this PR aim to accomplish?:**
- Replace lighttpd and PHP with the new API and web interface.
- Remove now-unnecessary files from the repo (lighttpd configs, PHP block page, etc)
- Delegate dnsmasq config generation to the API.

**How does this PR accomplish the above?:**
- The lighttpd- and PHP-specific code and files are removed from the repository.
- The API is added in a similar way as FTL, including support for updating and checking out different branches.
- The dnsmasq config generation code runs an API command to generate the config instead of manually modifying it as it did before.

**What documentation changes (if any) are needed to support this PR?:**
Remove documentation on lighttpd and PHP, replace with API and new web interface documentation.

**TODO Items**
These are some general items, but check the [API and web interface board](https://github.com/orgs/pi-hole/projects/2) for more items.
- [ ] Add an upgrade path from the current Pi-hole version (with lighttpd and PHP) to this one.
- [x] Add support for not hosting the web interface (`INSTALL_WEB_INTERFACE=false`)
- [ ] Write lots of documentation about the API and web interface
- [ ] Start a beta period
- [x] Support CentOS. Currently, the API is compiled with too new of a glibc version. A fix is to change the build system to use CentOS for building. This would make sure any distro that's at least as new as CentOS can run the API.
